File: server/src/main/java/bgu/spl/net/impl/stomp/Backend/Frames/Frame.java
package bgu.spl.net.impl.stomp.Backend.Frames;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import bgu.spl.net.impl.stomp.Backend.StompFacade;

public abstract class Frame {

    protected static final StompFacade op = StompFacade.getInstance();

    protected final static String END_OF_FRAME = "^@";
    protected final static String NEW_LINE = "\n";
    protected final static String HEADER_DELIMITER = ":";

    protected enum StompCommand {
        CONNECT, CONNECTED, SEND, SUBSCRIBE, UNSUBSCRIBE, DISCONNECT, MESSAGE, RECEIPT, ERROR
    }

    protected static class HeaderLine{
        public final String headerName;
        public final String headerValue;

        public HeaderLine(String key, String value) {
            this.headerName = key;
            this.headerValue = value;
        }
    }

    protected Frame(List<HeaderLine> headers, String frameBody, StompCommand command){
        instantiateHeaders(headers);
        this.frameBody = frameBody;
        this.command = command;
    }

    protected StompCommand command;
    protected HeaderLine[] headers;
    protected String frameBody;


    //================================================================================================|
    //=================================== Main Methods  ==============================================|
    //================================================================================================|

    /**
     * Create a frame according to the command received
     * Returns
     */
    public static Frame parse(String messageToParse){
        String[] frameParameters = messageToParse.split(NEW_LINE); // split by new line
        StompCommand command = StompCommand.valueOf(frameParameters[0]); // parse command

        // parse headers
        List<HeaderLine> headers = new LinkedList<HeaderLine>();
        int frameParametersLine = 1;
        while (!frameParameters[frameParametersLine].trim().equals(""))
        {
            String[] header = frameParameters[frameParametersLine].split(HEADER_DELIMITER);
            headers.add(new HeaderLine(header[0], header[1]));
            frameParametersLine++;
        }
        frameParametersLine++;
        // parse body
        String frameBody = "";
        while (!frameParameters[frameParametersLine].trim().equals(END_OF_FRAME))
        {
            frameBody += frameParameters[frameParametersLine]+NEW_LINE;
            frameParametersLine++;
        }

        if (frameBody.equals("")) {
            frameBody = null;
        }

        return createFrame(command, headers, frameBody);
    }

    public String getFrameString(){
        String output = "";
        output += command + NEW_LINE;
        for (HeaderLine header : headers){
            output += header.headerName + HEADER_DELIMITER + header.headerValue + NEW_LINE;
        }
        output += NEW_LINE;
        if (frameBody != null){
            output += frameBody;
        }
        output += END_OF_FRAME;
        return output;
    }

    //================================================================================================|
    //==================================== Utility Methods ===========================================|
    //================================================================================================|

    protected String encapsulateBody (String body){
        return "Received frame was:\n====================\n" + body + "====================\n";
    }

    private void instantiateHeaders(List<HeaderLine> headers){
        this.headers = new HeaderLine[headers.size()];
        Iterator<HeaderLine> iter = headers.iterator();
        int i = 0;
        while (iter.hasNext()){
            this.headers[i] = iter.next();
            i++;
        }
    }

    private static Frame createFrame(StompCommand command, List<HeaderLine> headers, String frameBody){
        switch (command){
            case CONNECT:
                return new ConnectFrame(headers, frameBody);
            case CONNECTED:
                return new ConnectedFrame(headers, frameBody);
            case SEND:
                return new SendFrame(headers, frameBody);
            case SUBSCRIBE:
                return new SubscribeFrame(headers, frameBody);
            case UNSUBSCRIBE:
                return new UnsubscribeFrame(headers, frameBody);
            case DISCONNECT:
                return new DisconnectFrame(headers, frameBody);
            case MESSAGE:
                return new MessageFrame(headers, frameBody);
            case RECEIPT:
                return new ReceiptFrame(headers, frameBody);
            case ERROR:
                return new ErrorFrame(headers, frameBody);
            default:
                return null;
        }
    }
        
}

File: client/src/Frames/Frame.cpp
#include "Frame.h"
#include "ConnectFrame.h"
#include "ConnectedFrame.h"
#include "SendFrame.h"
#include "SubscribeFrame.h"
#include "UnsubscribeFrame.h"
#include "DisconnectFrame.h"
#include "MessageFrame.h"
#include "ReceiptFrame.h"
#include "ErrorFrame.h"

UserData& Frame::userData = UserData::getInstance();

Frame::Frame(StompCommand command, unordered_map<string, string> headers, string frameBody)
    : headers(headers), frameBody(frameBody), command(command) {}

Frame* Frame::parse(string messageToParse) {
    istringstream stream(messageToParse);
    string line;
    getline(stream, line);
    StompCommand command = parseCommand(line);

    unordered_map<string, string> headers;
    while (getline(stream, line) && line != "") {
        size_t delimiterPos = line.find(HEADER_DELIMITER);
        string headerName = line.substr(0, delimiterPos);
        string headerValue = line.substr(delimiterPos + 1);
        headers[headerName] = headerValue;
    }

    string frameBody;
    while (getline(stream, line) && line != string(1, END_OF_FRAME)) {
        frameBody += line + NEW_LINE;
    }

    if (frameBody.empty()) {
        frameBody = "";
    }

    return createFrame(command, headers, frameBody);
}

StompCommand Frame::parseCommand(string command) {
    if (command == "CONNECT") {
        return CONNECT;
    } else if (command == "CONNECTED") {
        return CONNECTED;
    } else if (command == "SEND") {
        return SEND;
    } else if (command == "SUBSCRIBE") {
        return SUBSCRIBE;
    } else if (command == "UNSUBSCRIBE") {
        return UNSUBSCRIBE;
    } else if (command == "DISCONNECT") {
        return DISCONNECT;
    } else if (command == "MESSAGE") {
        return MESSAGE;
    } else if (command == "RECEIPT") {
        return RECEIPT;
    } else if (command == "ERROR") {
        return ERROR;
    } else {
        throw "Invalid command";
    }
}

Frame* Frame::createFrame(StompCommand command, unordered_map<string, string> headers, string frameBody) {

    switch (command) {
        case CONNECT:
            return new ConnectFrame(command, headers, frameBody);
        case CONNECTED:
            return new ConnectedFrame(command, headers, frameBody);
        case SEND:
            return new SendFrame(command, headers, frameBody);
        case SUBSCRIBE:
            return new SubscribeFrame(command, headers, frameBody);
        case UNSUBSCRIBE:
            return new UnsubscribeFrame(command, headers, frameBody);
        case DISCONNECT:
            return new DisconnectFrame(command, headers, frameBody);
        case MESSAGE:
            return new MessageFrame(command, headers, frameBody);
        case RECEIPT:
            return new ReceiptFrame(command, headers, frameBody);
        case ERROR:
            return new ErrorFrame(command, headers, frameBody);
        default: return nullptr;
    }
}

string Frame::toString() {
    string output = "";
    output += command + NEW_LINE;
    for (const auto& header : headers) {
        output += header.first + HEADER_DELIMITER + header.second + NEW_LINE;
    }
    output += NEW_LINE;
    if (!frameBody.empty()) {
        output += frameBody;
    }
    output += END_OF_FRAME;
    return output;
}

StompCommand Frame::getCommand()
{
    return StompCommand();
}

unordered_map<string, string> Frame::getHeaders()
{
    return unordered_map<string, string>();
}
